lib/scholar/metrics/classification.ex

```diff
@@ -586,86 +589,174 @@ defmodule Scholar.Metrics.Classification do
+    recall = safe_division(true_positive, true_positive + false_negative)
+
+    per_class_fscore =
+      cond do
+        # Should only be +Inf
+        Nx.is_infinity(beta) ->
+          recall
+        beta == 0 ->
+          precision
+        true ->
+          beta2 = Nx.pow(beta, 2)
+          safe_division((1 + beta2) * precision * recall, beta2 * precision + recall)
+      end
 
-        recall = safe_division(true_positive, true_positive + false_negative)
+    case average do
+      :none ->
+        {precision, recall, per_class_fscore}
 
-        per_class_f1 = safe_division(2 * precision * recall, precision + recall)
+      :micro ->
+        {precision, recall, per_class_fscore}
 
-        case opts[:average] do
-          :none ->
-            per_class_f1
+      :macro ->
+        {precision, recall, Nx.mean(per_class_fscore)}
 
-          :macro ->
-            Nx.mean(per_class_f1)
+      :weighted ->
+        support = (y_true == Nx.iota({num_classes, 1})) |> Nx.sum(axes: [1])
 
-          :weighted ->
-            support = (y_true == Nx.iota({num_classes, 1})) |> Nx.sum(axes: [1])
+        per_class_fscore =
+          per_class_fscore * support
+          |> safe_division(Nx.sum(support))
+          |> Nx.sum()
 
-            safe_division(per_class_f1 * support, Nx.sum(support))
-            |> Nx.sum()
-        end
+        {precision, recall, per_class_fscore}
     end
   end
 
+  @doc """
+  Calculates F1 score given rank-1 tensors which represent
+  the expected (`y_true`) and predicted (`y_pred`) classes.
+
+  If all examples are true negatives, then the result is 0 to
+  avoid zero division.
+
+  ## Options
+
+  #{NimbleOptions.docs(@f1_score_schema)}
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: :u32)
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: :u32)
+      iex> Scholar.Metrics.Classification.f1_score(y_true, y_pred, num_classes: 3)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 0.6666666865348816, 0.4000000059604645]
+      >
+      iex> Scholar.Metrics.Classification.f1_score(y_true, y_pred, num_classes: 3, average: :macro)
+      #Nx.Tensor<
+        f32
+        0.5777778029441833
+      >
+      iex> Scholar.Metrics.Classification.f1_score(y_true, y_pred, num_classes: 3, average: :weighted)
+      #Nx.Tensor<
+        f32
+        0.6399999856948853
+      >
+      iex> Scholar.Metrics.Classification.f1_score(y_true, y_pred, num_classes: 3, average: :micro)
+      #Nx.Tensor<
+        f32
+        0.6000000238418579
+      >
+      iex> Scholar.Metrics.Classification.f1_score(Nx.tensor([1, 0, 1, 0]), Nx.tensor([0, 1, 0, 1]), num_classes: 2, average: :none)
+      #Nx.Tensor<
+        f32[2]
+        [0.0, 0.0]
+      >
+  """
+  deftransform f1_score(y_true, y_pred, opts \\ []) do
+    fbeta_score_n(y_true, y_pred, Nx.tensor(1), NimbleOptions.validate!(opts, @f1_score_schema))
```

Comment:
Not sure if `Nx.tensor(1)` could be marginally improved by changing to `Nx.tensor(1, type: :u8)` or if, since this is being passed to a `defn` context if the tensor creation is necessary (currently not enforced). From testing, it doesn't seems to make a difference in results.

Reply:
You could just pass 1 directly and let it be converted. Though given that the output type is floating point, any integer type tensor will work